doc/dev/crimson/poseidonstore.rst

```diff
@@ -0,0 +1,457 @@
+    latest meta data
+
+* Flush (WAL --> Data partition)
+  - Flush WAL entries which have committed. There are two conditions
+    (1. the size of WAL is close to full, 2. a signal to flush).
+    We can mitigate the overhead of frequent flush via batching processing, but it leads to
+    delaying completion.
+
+
+Crash consistency
+------------------
+* Large case
+  - 1. Crash occurs right after writing Data blocks
+    - Data partition --> | Data blocks |
+    - We don't need to care this case. Data is not alloacted yet in reality. The blocks will be reused.
+  - 2. Crash occurs right after WAL 
+    - Data partition --> | Data blocks |
+    - WAL --> | TxBegin A | Log Entry | TxEnd A |
+    - Write procedure is completed, so there is no data loss or inconsistent state
+
+* Small case
+  - 1. Crash occurs right after writing WAL
+    - WAL --> | TxBegin A | Log Entry| TxEnd A |
+    - All data has been written
+
+
+Comparison
+----------
+- Best case (pre-allocation)
+  - Only need writes on both WAL and Data partition without updating object metadata (for the location).
+- Worst case 
+  - At least three writes are required additionally on WAL, object metadata, and data blocks.
+  - If the flush from WAL to the data parition occurs frequently, b+tree onode structure needs to be update
+    in many times. To minimize such overhead, we can make use of batch processing to minimize the update on the btree
+    (the data related to the object has a locality because it will have the same parent node, so updates can be minimized)
+
+- WAL needs to be flushed if the WAL is close to full or a signal to flush.
+  - The premise behind this design is OSD can manage the lateset metadta as a single copy. So,
+    appended entires are not to be read
+- Either best of worst case does not produce severe I/O amplification (it produce I/Os, but I/O rate is constant) 
+  unlike LSM-tree DB (proposed design is similar to LSM-tree which has only level-0)
+  
+
+Detailed Design 
+===============
+
+- Onode lookup
+  - Radix tree
+  Our design is entirely based on the prefix tree. Ceph already makes use of the chracteristic of OID's prefix to split or search
+  the OID (e.g., pool id + hash + oid). So, the prefix tree fits well to store or search the object. Our scheme is designed 
+  to lookup the prefix tree efficiently.
+
+  - Sharded partition
+  A few bits at the begining of OID determine a sharded partition where the object is located.
+   +-------------------------------+-------------------------------+-------------------------------+
+   | sharded partition 1 (bits:00) | sharded partition 2 (bits:01) | sharded partition 3 (bits:10) |
+   +-------------------------------+-------------------------------+-------------------------------+ 
+
+  - Ondisk onode
+    
+  stuct onode {
+    radix_tree childs;
+    radix_tree left, right;
+    radix_tree parent_node;
+    extent_tree block_maps;
+    clone_list clones; // tree list
```

Comment:
I don't understand this part.  Can you discuss how this maps onto the clone() and clone_range() operations?

Reply:
See below